local-volume/provisioner/pkg/monitor/monitor.go

```diff
@@ -0,0 +1,333 @@
+		hasRun:        false,
+		hasRunLock:    &sync.Mutex{},
+		volumeMux:     &sync.Mutex{},
+		volumeMap:     make(map[string]*v1.PersistentVolume),
+	}
+
+	monitor.volumeLW = &cache.ListWatch{
+		ListFunc: func(options metav1.ListOptions) (runtime.Object, error) {
+			return config.Client.CoreV1().PersistentVolumes().List(options)
+		},
+		WatchFunc: func(options metav1.ListOptions) (watch.Interface, error) {
+			return config.Client.CoreV1().PersistentVolumes().Watch(options)
+		},
+	}
+	_, monitor.volumeController = cache.NewInformer(
+		monitor.volumeLW,
+		&v1.PersistentVolume{},
+		monitor.resyncPeriod,
+		cache.ResourceEventHandlerFuncs{
+			AddFunc:    monitor.addVolume,
+			UpdateFunc: monitor.updateVolume,
+			DeleteFunc: monitor.deleteVolume,
+		},
+	)
+
+	// fill map at first with data from ETCD
+	monitor.flushFromETCDFirst()
+
+	return monitor
+}
+
+// flushFromETCDFirst fill map with data from etcd at first
+func (monitor *Monitor) flushFromETCDFirst() error {
+	pvs, err := monitor.Client.CoreV1().PersistentVolumes().List(metav1.ListOptions{})
+	if err != nil {
+		return err
+	}
+	if len(pvs.Items) == 0 {
+		glog.Infof("no pv in ETCD at first")
+		return nil
+	}
+	monitor.volumeMux.Lock()
+	defer monitor.volumeMux.Unlock()
+	for _, pv := range pvs.Items {
+		monitor.volumeMap[pv.Name] = &pv
+	}
+	return nil
+}
+
+func (monitor *Monitor) addVolume(obj interface{}) {
+	volume, ok := obj.(*v1.PersistentVolume)
+	if !ok {
+		glog.Errorf("Expected PersistentVolume but handler received %#v", obj)
+		return
+	}
+
+	monitor.volumeMux.Lock()
+	defer monitor.volumeMux.Unlock()
+	monitor.volumeMap[volume.Name] = volume
+
+}
+
+func (monitor *Monitor) updateVolume(oldObj, newObj interface{}) {
+	volume, ok := newObj.(*v1.PersistentVolume)
+	if !ok {
+		glog.Errorf("Expected PersistentVolume but handler received %#v", newObj)
+		return
+	}
+
+	monitor.volumeMux.Lock()
+	defer monitor.volumeMux.Unlock()
+	monitor.volumeMap[volume.Name] = volume
+}
+
+func (monitor *Monitor) deleteVolume(obj interface{}) {
+	volume, ok := obj.(*v1.PersistentVolume)
+	if !ok {
+		glog.Errorf("Expected PersistentVolume but handler received %#v", obj)
+		return
+	}
+
+	monitor.volumeMux.Lock()
+	defer monitor.volumeMux.Unlock()
+	delete(monitor.volumeMap, volume.Name)
+
+}
+
+// Run starts all of this controller's control loops
+func (monitor *Monitor) Run(stopCh <-chan struct{}) {
+	glog.Infof("Starting monitor controller %s!", string(monitor.RuntimeConfig.Name))
+	monitor.hasRunLock.Lock()
+	monitor.hasRun = true
+	monitor.hasRunLock.Unlock()
+	go monitor.volumeController.Run(stopCh)
+
+	go monitor.MonitorLocalVolumes()
+	<-stopCh
+}
+
+// HasRun returns whether the volume controller has Run
+func (monitor *Monitor) HasRun() bool {
+	monitor.hasRunLock.Lock()
+	defer monitor.hasRunLock.Unlock()
+	return monitor.hasRun
+}
+
+func (monitor *Monitor) MonitorLocalVolumes() {
+	for {
+		if monitor.HasRun() {
+			for _, pv := range monitor.volumeMap {
```

Comment:
`volumeMap` was protected by a mutex?

Reply:
yeah, and i'm thinking whether we can reuse the volumecache,  any problem ?

Reply:
I think there should be a mutex check while looping this map.

By `volumecache `, you mean `Store` returned by `Informer`?

Reply:
No, not `store`,  volume cache in `RuntimeConfig `
And this is a `get` opeation, will not change the map,  maybe do not need to add mutex

Reply:
Maps are not safe for concurrent use, no matter read or write operations.

Ref: https://blog.golang.org/go-maps-in-action Check the `Concurrency ` Part.

Reply:
And, why don't we use the `Store.List` directly?

Reply:
`Store` will store all PVs there and I'm thinking just focus on PVs(local storage) belonging to that node.

Reply:
btw: PVs may be created manually. We also need to take this into consideration too.